src/build/specialize_types/IR.zig

```diff
@@ -125,129 +147,157 @@ pub const Expr = union(enum) {
 
     compiler_bug: Problem.Compiler,
 
-    /// todo
+    /// List of Expr
     pub const List = collections.SafeList(@This());
-    /// todo
+    /// Index into an Expr.List
     pub const Idx = List.Idx;
-    /// todo
+    /// Sublist reference of Expr.List
     pub const Slice = List.Slice;
-    /// todo
+    /// Sublist reference into Expr.List that is guaranteed to be non-empty
     pub const NonEmptySlice = List.NonEmptySlice;
-    /// todo
+    /// ??? Type annotated expression e.g. `1: U64` ?
     pub const Typed = struct {
         expr: Expr.Idx,
         type: Type.Idx,
-        /// todo
+        /// List of Expr.Typed
         pub const List = collections.SafeMultiList(@This());
-        /// todo
+        /// Sublist reference into Expr.Typed.List
         pub const Slice = Typed.List.Slice;
     };
 };
 
 /// A definition, e.g. `x = foo`
 pub const Def = struct {
+    /// ??? Is pattern for what's to the left of `=`?
     pattern: Pattern.Idx,
-    /// Named variables in the pattern, e.g. `a` in `Ok a ->`
+    /// Named variables in the pattern, e.g. `a` in `Ok a ->` ??? A when branch is not a Def right?
     pattern_vars: TypedIdent.Slice,
+    /// expression to the right of `=`
     expr: Expr.Idx,
+    /// type of the expression to the right of `=`
     expr_type: Type.Idx,
-    /// todo
+    /// List of Def
     pub const List = collections.SafeMultiList(@This());
-    /// todo
+    /// Sublist reference into Def.List
     pub const Slice = List.Slice;
 };
 
-/// todo
+/// Branch of a `when` expression, e.g. `Green -> "green"`
 pub const WhenBranch = struct {
     /// The pattern(s) to match the value against
     patterns: Pattern.NonEmptySlice,
     /// A boolean expression that must be true for this branch to be taken
     guard: ?Expr.Idx,
     /// The expression to produce if the pattern matches
     value: Expr.Idx,
-    /// todo
+    /// List of WhenBranch
     pub const List = collections.SafeList(@This());
-    /// todo
+    /// Reference to non-empty sublist into WhenBranch.List
     pub const NonEmptySlice = List.NonEmptySlice;
 };
 
-/// todo
+/// e.g. `|x| x * 2`
 pub const Function = struct {
     args: Pattern.Slice,
     return_type: Type.Idx,
+    /// ??? Is expr the body of the function?
     expr: Expr.Idx,
 };
 
-/// todo
+/// ??? For Record destructuring, e.g. `{ x, y } = coords(point)`
 pub const StructDestruct = struct {
+    /// ???
     ident: Ident.Idx,
+    /// ???
     field: Ident.Idx,
     kind: Kind,
-    /// todo
+    /// ???
     pub const Kind = union(enum) {
+        /// ???
         required,
+        /// ???
         guard: Pattern.Typed,
     };
```

Comment:
This is close, but more specifically this is for each field in a destruct. In the following example:

```elm
match data {
    { x, y: 123 } -> ...
}
```

The field `x` is required and saved, meaning we'd have one `StructDestruct` with `ident = x`, `field = x`, and `kind = .required`.

The `y` field is just there for checking that its value is `123`, but the value is discarded, meaning we'd have a second `StructDestruct` with `ident = y`, `field = y`, and `kind = .{ .guard = .{ .num_literal = 123 } }`.